### PR TITLE
Razor Pages Filters

### LINKED ISSUE #5754: Document IPageFilter and filter methods on PageModel
`@pranavkm` wrote:   `IActionFilter`: https://github.com/aspnet/Home/issues/2981#issuecomment-374706661

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #5754 
[Review URL](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/razor-pages/filter?branch=pr-en-us-5835)

Clean up of [Filters in ASP.NET Core](https://review.docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?branch=pr-en-us-5835)

### CODE REVIEW

File: aspnetcore/mvc/razor-pages/filter.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+The following code implements the synchronous `IPageFilter`:
+
+[!code-csharp[Main](filter/sample/PageFilter/Filters/SamplePageFilter.cs?name=snippet1)]
+
+The following code enables the `SamplePageFilter`:
+
+[!code-csharp[Main](filter/sample/PageFilter/StartupSync.cs?name=snippet2&highlight=11)]
+
+## Implement Razor Page filters by overriding page handlers
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/filter.md]: overriding filter methods (not handlers)

**Comment by Rick-Anderson** [comment]
@pranavkm  can you send me the code for a filter attribute?

**Comment by pranavkm** [comment]
You could use the same `ResultFilter` example as https://review.docs.microsoft.com/en-us/aspnet/core/mvc/controllers/filters?branch=pr-en-us-5835#built-in-filter-attributes.

**Comment by Rick-Anderson** [comment]
@pranavkm can all the following be used with Razor Pages?
I added a sample with `ResultFilterAttribute`.
```
ActionFilterAttribute
ExceptionFilterAttribute
ResultFilterAttribute
FormatFilterAttribute
ServiceFilterAttribute
TypeFilterAttribute
```

It should be ready to review. I'll update the sample to  `2.1.0-preview2-final` when it's available.

**Comment by Rick-Anderson** [comment]
@pranavkm  ready for review. Do all the `XyzFilterAttributes` I listed work with Razor Pages?

**Comment by Rick-Anderson** [comment]
@fmaeseele please review.

**Comment by Rick-Anderson** [comment]
@pranavkm Can you take another look. It should be ready.

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Filters/AddHeaderAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        private readonly string _name;
+        private readonly string _value;
+
+        public AddHeaderAttribute (string name, string value)
+        {
+            _name = name;
+            _value = value;
+        }
+
+        public override async Task OnResultExecutionAsync(
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Filters/AddHeaderAttribute.cs]: Could you change this to `public override void `OnResultExecuting(ResultExecutingContext context)`? Would be simpler than the async overload.

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Pages/subFolder/About2.cshtml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿@page
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Pages/subFolder/About2.cshtml]: Is this intentional? The folder name \ file name seem a bit suspect.

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Program.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    public class Program
+    {
+        public static void Main(string[] args)
+        {
+            CreateWebHostBuilder(args).Build().Run();
+        }
+
+        public static IWebHostBuilder CreateWebHostBuilder(string[] args) =>
+            WebHost.CreateDefaultBuilder(args)
+             //      .UseStartup<StartupSync>();
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Program.cs]: Cleanup

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Startup2.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
+using Microsoft.AspNetCore.Builder;
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Startup2.cs]: ?

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Startup2.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+        #region snippet2
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddMvc()
+               .AddRazorPagesOptions(options =>
+               {
+                   options.Conventions.AddFolderApplicationModelConvention(
+                       "/subFolder",
+                       model => model.Filters.Add(new SampleAsyncPageFilter(_logger)));
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Startup2.cs]: Is there value in showing a sync and async filter being added globally? I'd pick one as long as we show a sample of a sync filter and an async filter being used.

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/wwwroot/css/site.css [hunk 7]
```diff
@@ -0,0 +1,37 @@
+/* Please see documentation at https://docs.microsoft.com/aspnet/core/client-side/bundling-and-minification
```
Comment (pranavkm) [inline aspnetcore/mvc/razor-pages/filter/sample/PageFilter/wwwroot/css/site.css]: Out of curiosity, do we always include all the css \ scripts with these samples?

File: aspnetcore/toc.md [hunk 8]
```diff
@@ -246,7 +249,6 @@
 ### [Build Docker images](/dotnet/articles/core/docker/building-net-docker-images)
 ### [Visual Studio Tools for Docker](xref:host-and-deploy/docker/visual-studio-tools-for-docker)
 ### [Publish to a Docker image](https://azure.microsoft.com/documentation/articles/vs-azure-tools-docker-hosting-web-apps-in-docker/)
-## [Proxy and load balancer configuration](xref:host-and-deploy/proxy-load-balancer)
```
Comment (pranavkm) [inline aspnetcore/toc.md]: ?

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Pages/subFolder/About2.cshtml (hunk 3 above)
Comment (Rick-Anderson) [reply aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Pages/subFolder/About2.cshtml]: I deleted about2. I don't need it to show:
```
services.AddMvc()
   .AddRazorPagesOptions(options =>
   {
       options.Conventions.AddFolderApplicationModelConvention(
           "/subFolder",
           model => model.Filters.Add(new SampleAsyncPageFilter(_logger)));
   });
```

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Program.cs (hunk 4 above)
Comment (Rick-Anderson) [reply aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Program.cs]: I formatted the commented out code but I need all 3 startup class references to test the 3 startup classes - depending on what I want to show.

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Startup2.cs (hunk 5 above)
Comment (Rick-Anderson) [reply aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Startup2.cs]: image

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Startup2.cs (hunk 6 above)
Comment (Rick-Anderson) [reply aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Startup2.cs]: That sample is to show `AddFolderApplicationModelConvention`
In the rendered document, this async filter is shown in the previous snippet.

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/wwwroot/css/site.css (hunk 7 above)
Comment (Rick-Anderson) [reply aspnetcore/mvc/razor-pages/filter/sample/PageFilter/wwwroot/css/site.css]: I lost that battle long ago. I used to exclude them with *.gitignore* and require you to manually override the ignore file. Now I ask folks to manually exclude them which is usually forgotten.

File: aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Program.cs (hunk 4 above)
Comment (pranavkm) [reply aspnetcore/mvc/razor-pages/filter/sample/PageFilter/Program.cs]: Okies

**Pull request merged**